toqito/matrices/operator_sinkhorn.py

```diff
@@ -0,0 +1,190 @@
+
+import numpy as np
+import scipy as sp
+
+from toqito.channels.partial_trace import partial_trace
+from toqito.matrix_props.is_square import is_square
+
+
+def operator_sinkhorn(
+    rho: np.ndarray, dim: list[int] = None, tol: float = np.sqrt(np.finfo(float).eps), max_iterations: int = 10_000
+) -> tuple[np.ndarray, list[np.ndarray]]:
+    r"""Perform operator Sinkhorn iteration (iterative sinkhorn algorithm) on a quantum system's density matrix.
+
+    This function is adapted from QETLAB. :cite:`QETLAB_link`.
+
+    This function relies on Sinkhorn's theorem :cite:`Sinkhorn_1964_Relationship` which states "for any
+    positive-definite square matrix, there exist diagonal matrices :math:`D_1` and :math:`D_2` such that
+    :math:`D_1 \, \cdot A \, \cdot D_2` is doubly stochastic.
+
+    The iterative Sinkhorn algorithm alternately rescales the input density matrix of the quantum system along each
+    subsystem so that its marginals become uniform (i.e., partial traces or row/column sums become proportional to
+    identity), until convergence.
+
+    Upon convergence we end up with a density matrix `sigma` (:math:`\sigma`), which is locally equivalent to the given
+    multipartite density matrix `rho` (:math:`\rho`), but having maximally mixed subsystems. The algorithm will also
+    return the list of Filtering Operations which can be directly applied on :math:`\rho` to arrive at :math:`\sigma`.
+    Such converted forms of density matrices are useful to analyse entangled systems and to study effects of operations
+    on each subsystems. (as discussed in :cite:`Gurvits_2004_Classical`)
+
+    Examples
+    ==========
+
+    To obtain a locally normalized density matrix of a random 2-qubit bipartite state using operator Sinkhorn:
+
+    ..jupyter-execute::
+        import numpy as np
+        from toqito.matrices import operator_sinkhorn
+        from toqito.rand import random_density_matrix
+
+        rho_random = random_density_matrix(4, seed=42)
+
+        print(rho_random)
+
+    :code:`operator_sinkhorn` returns the result density matrix along with the operations list :code:`F`.
+    :code:`sigma` (:math:`\sigma`) has all of its (single-party) reduced density matrices proportional
+    to the identity, while satisfying
+
+    .. math::
+        \sigma \, = \, F \, \cdot \, \rho \, \cdot \, F^{\dagger}
+
+    In other words, F contains invertible local operations that demonstrate
+    that :code:`rho` and :code:`sigma` are locally equivalent.
+
+    ..jupyter-execute::
+        sigma, F = operator_sinkhorn(rho=rho_random, dim=[2, 2])
+
+        print(sigma)
+
+        print(len(F))
+
+    F here will be the list of 2 local filtering operators, which are
+
+    ..jupyter-execute::
+        print(F[0])
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :param rho: input density matrix of a multipartite system
+    :param dim: `None` list containing dimensions of each subsystem.
```

Comment:
The wording of this is a little weird. The default argument is `None`, but it could also be a list as you describe. 
```suggestion
    :param dim: `None` list containing dimensions of each subsystem.
```

Reply:
@vprusso Edited it now. Please let me know if its good to go.